src/codegen.cpp

```diff
@@ -6610,6 +6615,9 @@ static std::unique_ptr<Module> emit_function(
             }
             if (!jl_is_uniontype(phiType) || !TindexN) {
                 if (VN) {
+                    // XXX: this code assumes that `val` is of type `phiType` statically,
```

Comment:
As I mentioned in person, I do want this invariant to hold statically. I'm working on a PR to add that to the verifier and fix the cases that currently violate it.

Reply:
Perhaps. Currently the type-system is incapable of maintaining your desired invariant however (it's most common to observe this during IPO transforms such as inlining, although rare to come across such an example and thus not simple to construct from first principles a reliable example).